README
                         PulseAudio Volume Meter 0.9.3

   Copyright 2004-2007 Lennart Poettering <mzihzrgre (at) 0pointer (dot)
   de>
     * [1]License
     * [2]News
     * [3]Overview
     * [4]Current Status
     * [5]Documentation
     * [6]Requirements
     * [7]Installation
     * [8]Acknowledgements
     * [9]Download

License

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 2 of the License, or (at your
   option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   675 Mass Ave, Cambridge, MA 02139, USA.

News

   Tue Oct 30 2007:

   [10]Version 0.9.3 released; lower refresh rate; show pretty channel
   names; other fixes

   Sat Jul 8 2006:

   [11]Version 0.9.2 released; compatibility with [12]PulseAudio 0.9.2

   Sat May 27 2006:

   [13]Version 0.9.0 released; compatibility with Polypaudio 0.9.0

   Fri Apr 21 2006:

   [14]Version 0.8 released; compatibility with Polypaudio 0.8

   Sun Nov 21 2004:

   [15]Version 0.3 released; compatibility with Polypaudio 0.7; decaying
   of signal levels

   Thu Oct 28 2004:

   [16]Version 0.3 released; compatibility with Polypaudio 0.6

   Mon Sep 20 2004:

   [17]Version 0.2 released; compatibility with Polypaudio 0.5

   Wed Sep 9 2004:

   [18]Version 0.1 released

Overview

   PulseAudio Volume Meter (pavumeter) is a simple GTK volume meter for
   the [19]PulseAudio sound server.

   Everybody loves [20]screenshots.

Current Status

   Works perfectly.

Documentation

   There is not much to say. Just run pavumeter and see yourself. You may
   specify the source to monitor on the command line. If ommited the
   daemon's default source is selected.

Requirements

   Currently, pavumeter is tested on Linux only.

   pavumeter was developed and tested on Debian GNU/Linux "testing" from
   September 2004, it should work on most other Linux distributions (and
   maybe Unix versions) since it uses GNU autoconf for source code
   configuration.

   pavumeter requires [21]gtkmm installed.

   Obviously pavumeter requires an installation of [22]PulseAudio (version
   0.9.2).

Installation

   As this package is made with the GNU autotools you should run
   ./configure inside the distribution directory for configuring the
   source tree. After that you should run make for compilation and make
   install (as root) for installation of pavumeter.

Acknowledgements

   None so far.

Download

   The newest release is always available from
   [23]http://0pointer.de/lennart/projects/pavumeter/

   The current release is [24]0.9.3

   Get pavumeter's development sources from the [25]Subversion
   [26]repository ([27]viewcvs):
svn checkout svn://svn.0pointer.net/pavumeter/trunk pavumeter

   If you want to be notified whenever I release a new version of this
   software use the subscription feature of [28]Freshmeat.
     __________________________________________________________________


    Lennart Poettering <mzihzrgre (at) 0pointer (dot) de>, Oct 2007

   $Id: README.html.in 54 2007-10-30 16:43:04Z lennart $

References

   1. README#license
   2. README#news
   3. README#overview
   4. README#status
   5. README#documentation
   6. README#requirements
   7. README#installation
   8. README#acks
   9. README#download
  10. http://0pointer.de/lennart/projects/pavumeter/pavumeter-0.9.3.tar.gz
  11. http://0pointer.de/lennart/projects/pavumeter/pavumeter-0.9.2.tar.gz
  12. http://pulseaudio.org/
  13. http://0pointer.de/lennart/projects/pavumeter/pavumeter-0.9.0.tar.gz
  14. http://0pointer.de/lennart/projects/pavumeter/pavumeter-0.8.tar.gz
  15. http://0pointer.de/lennart/projects/pavumeter/pavumeter-0.4.tar.gz
  16. http://0pointer.de/lennart/projects/pavumeter/pavumeter-0.3.tar.gz
  17. http://0pointer.de/lennart/projects/pavumeter/pavumeter-0.2.tar.gz
  18. http://0pointer.de/lennart/projects/pavumeter/pavumeter-0.1.tar.gz
  19. http://0pointer.de/lennart/projects/pulseaudio/
  20. http://0pointer.de/lennart/projects/pavumeter//screenshot.png
  21. http://gtkmm.sf.net/
  22. http://pulseaudio.org/
  23. http://0pointer.de/lennart/projects/pavumeter/
  24. http://0pointer.de/lennart/projects/pavumeter/pavumeter-0.9.3.tar.gz
  25. http://subversion.tigris.org/
  26. svn://svn.0pointer.net/pavumeter
  27. http://0pointer.net/cgi-bin/viewcvs.cgi/?root=pavumeter
  28. http://freshmeat.net/projects/pavumeter/


ChangeLog
2007-10-30 16:43  lennart

	* Makefile.am, doc/README.html.in: prepare next release

2007-09-06 21:50  lennart

	* configure.ac, src/vumeter.cc: - lower refresh rate to 25 fps
	  - show pretty channel names
	  - show device description instead of device name

2007-08-15 23:49  lennart

	* Makefile.am, configure.ac: bump version and add fedora-snapshot
	  target

2007-08-07 00:21  lennart

	* src/Makefile.am, src/pavumeter-record.desktop,
	  src/pavumeter.desktop: add seperate desktop file for capture

2007-08-06 22:49  lennart

	* src/vumeter.cc: remove full stop

2007-08-06 17:37  lennart

	* src/vumeter.cc: Handle parameters properly

2007-08-06 15:54  lennart

	* src/vumeter.cc: make sure to use black color for subtitle to be
	  compatible with all themes

2007-05-23 23:34  lennart

	* LICENSE, autogen.sh, src/vumeter.cc: Fix license mess: change
	  everything to GPL

2006-07-24 20:17  lennart

	* autogen.sh: Make jhbuild happy by adding a file autgen.sh

2006-07-08 13:56  lennart

	* doc/README.html.in: prepare release 0.9.2

2006-06-21 14:12  lennart

	* src/vumeter.cc: include config.h properly

2006-06-20 23:59  lennart

	* configure.ac: bump requied pa version

2006-06-20 22:28  lennart

	* doc/README.html.in, src/vumeter.cc: s/polyp/pulse/ in README

2006-06-20 22:17  lennart

	* Makefile.am, bootstrap.sh, configure.ac, doc/Makefile.am,
	  doc/style.css, src/Makefile.am, src/pavumeter.desktop,
	  src/vumeter.cc: s/polyp/pulse/g

2006-06-02 21:09  lennart

	* src/vumeter.cc: do not show the polypaudio error code if we
	  terminate ourselves due to a reason that is unrelated to
	  polypaudio

2006-06-02 20:53  lennart

	* src/Makefile.am: actually ship the .desktop file

2006-06-02 17:05  ossman

	* src/vumeter.cc: Throw up a dialog for critical errors as stderr
	  might not be visible.

2006-05-31 12:33  ossman

	* src/Makefile.am, src/pavumeter.desktop: Add .desktop file.

2006-05-28 15:44  lennart

	* src/vumeter.cc: set icon for window, add icon to window title

2006-05-26 23:43  lennart

	* configure.ac: bump version

2006-05-26 23:36  lennart

	* doc/README.html.in: doc update

2006-05-13 23:26  lennart

	* configure.ac: bump polypaudio release

2006-04-26 19:17  lennart

	* src/vumeter.cc: compat with 64bit

2006-04-21 20:45  lennart

	* Makefile.am: build system updates

2006-04-21 20:39  lennart

	* bootstrap.sh, configure.ac, doc/README.html.in, doc/style.css:
	  doc update

2006-04-07 21:54  lennart

	* src/vumeter.cc: * update for newest polypaudio API
	  * show channel names from channel map

2006-04-07 21:53  lennart

	* configure.ac: update build system a little for polypaudio 0.8

2005-01-08 23:59  lennart

	* configure.ac, src/vumeter.cc: update for polypaudio 0.8

2004-12-03 15:46  lennart

	* doc/README.html.in: readme update

2004-11-21 18:49  lennart

	* configure.ac: bump version numbers

2004-11-17 03:10  lennart

	* src/vumeter.cc: * implement level decaying

2004-10-27 22:43  lennart

	* configure.ac, doc/README.html.in: prepare next release

2004-10-27 01:14  lennart

	* src/vumeter.cc: update to new API

2004-09-29 20:14  lennart

	* configure.ac, src/vumeter.cc: update to polypaudio 0.6

2004-09-20 21:53  lennart

	* doc/README.html.in: prepare for dist

2004-09-16 22:03  lennart

	* src/vumeter.cc: Don't ignore $POLYP_SOURCE
	  memory leak fix

2004-09-16 00:03  lennart

	* configure.ac, src/vumeter.cc: update for new API
	  show logarthmic levels

2004-09-12 13:17  lennart

	* src: ignore some more

2004-09-12 13:16  lennart

	* configure.ac, src/Makefile: update to polypaudio 0.5

2004-09-08 00:12  lennart

	* doc/README.html.in: typo fix in homepage

2004-09-07 23:54  lennart

	* ., LICENSE, Makefile.am: prepare for release

2004-09-07 23:49  lennart

	* Makefile.am, doc/README.html.in: do missing paman to pavumeter
	  replacements

2004-09-07 23:46  lennart

	* ., Makefile.am, bootstrap.sh, configure.ac, doc, doc/Makefile.am,
	  doc/README.html.in, doc/screenshot.png, doc/style.css, src,
	  src/Makefile, src/Makefile.am: autotoolize

2004-09-07 14:50  lennart

	* src/vumeter.cc: add a nifty title bar

2004-09-07 14:36  lennart

	* src/vumeter.cc: add support for latencies

2004-09-06 21:55  lennart

	* src, src/Makefile, src/vumeter.cc: make it work

2004-09-06 17:50  lennart

	* src, src/Makefile, src/vumeter.cc: initial commit

2004-09-06 17:48  lennart

	* .: Create trunk directory



bootstrap.sh
#!/bin/bash
# $Id: bootstrap.sh 40 2006-06-20 22:17:24Z lennart $

# This file is part of PulseAudio Volume Meter.
#
# PulseAudio Volume Meter is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# PulseAudio Volume Meter is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with PulseAudio Volume Meter; if not, write to the Free Software Foundation,
# Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.

VERSION=1.9

run_versioned() {
    local P
    local V

    V=$(echo "$2" | sed -e 's,\.,,g')
    
    if [ -e "`which $1$V`" ] ; then
    	P="$1$V" 
    else
	if [ -e "`which $1-$2`" ] ; then
	    P="$1-$2" 
	else
	    P="$1"
	fi
    fi

    shift 2
    "$P" "$@"
}

set -ex

if [ "x$1" = "xam" ] ; then
    run_versioned automake "$VERSION" -a -c --foreign
    ./config.status
else 
    rm -rf autom4te.cache
    rm -f config.cache

    run_versioned aclocal "$VERSION"
    run_versioned autoconf 2.59 -Wall
    run_versioned autoheader 2.59
    run_versioned automake "$VERSION" -a -c --foreign

    CFLAGS="-g -O0" ./configure --sysconfdir=/etc "$@"

    make clean
fi


src/vumeter.cc
/* $Id: vumeter.cc 53 2007-09-06 21:50:05Z lennart $ */

/***
  This file is part of pavumeter.

  pavumeter is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published
  by the Free Software Foundation; either version 2 of the License,
  or (at your option) any later version.

  pavumeter is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with pavumeter; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA.
***/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <signal.h>
#include <math.h>

#include <gtkmm.h>
#include <gtk/gtk.h>

#include <pulse/pulseaudio.h>
#include <pulse/glib-mainloop.h>

//#define LOGARITHMIC 1
#define DECIBEL 1

class MainWindow : public Gtk::Window {

public:
    MainWindow(const pa_channel_map &map, const char *source_name, const char *description);
    virtual ~MainWindow();

protected:

    class ChannelInfo {
    public:
        ChannelInfo(MainWindow &w, const Glib::ustring &l);
        Gtk::Label *label;
        Gtk::ProgressBar *progress;
    };

    Gtk::VBox vbox, titleVBox;
    Gtk::HBox titleHBox;
    Gtk::Table table;
    std::vector<ChannelInfo*> channels;
    Gtk::Image image;
    Gtk::Label titleLabel;
    Gtk::Label subtitleLabel;
    Gtk::HSeparator separator;
    Gtk::EventBox eventBox;

    float *levels;

    virtual void addChannel(const Glib::ustring &l);
    virtual bool on_delete_event(GdkEventAny* e);
    virtual bool on_display_timeout();
    virtual bool on_calc_timeout();
    virtual void decayLevels();

    sigc::connection display_timeout_signal_connection;
    sigc::connection calc_timeout_signal_connection;
    pa_usec_t latency;

    class LevelInfo {
    public:
        LevelInfo(float *levels, pa_usec_t l);
        virtual ~LevelInfo();
        bool elapsed();

        struct timeval tv;
        float *levels;
    };

    std::deque<LevelInfo *> levelQueue;

public:
    virtual void pushData(const float *d, unsigned l);
    virtual void showLevels(const LevelInfo& i);
    virtual void updateLatency(pa_usec_t l);
};

MainWindow::MainWindow(const pa_channel_map &map, const char *, const char *description) :
    Gtk::Window(),
    table(1, 2),
    latency(0) {

    char t[256];
    int n;

    set_title("PulseAudio Volume Meter");

    gtk_window_set_icon_name(GTK_WINDOW(gobj()), "audio-input-microphone");

    add(vbox);

    Gdk::Color c("white");
    eventBox.modify_bg(Gtk::STATE_NORMAL, c);

    vbox.pack_start(eventBox, false, false);

    image.set_from_icon_name("audio-input-microphone", Gtk::ICON_SIZE_DIALOG);

    eventBox.add(titleHBox);
    titleHBox.pack_start(image, false, false);
    titleHBox.pack_end(titleVBox, true, true);
    titleHBox.set_border_width(12);
    titleHBox.set_spacing(12);

    titleVBox.add(titleLabel);
    titleVBox.add(subtitleLabel);
    titleVBox.set_spacing(6);

    titleLabel.set_markup("<span size=\"18000\" color=\"black\"><b>PulseAudio Volume Meter</b></span>");
    titleLabel.set_alignment(0, 1);
    snprintf(t, sizeof(t), "<span color=\"black\">Showing signal levels of <b>%s</b></span>", description);
    subtitleLabel.set_markup(t);
    subtitleLabel.set_alignment(0, 0);

    vbox.pack_start(separator, false, false);

    table.set_border_width(12);
    table.set_row_spacings(6);
    table.set_col_spacings(12);
    vbox.pack_start(table, true, true);

    for (n = 0; n < map.channels; n++) {
        snprintf(t, sizeof(t), "<b>%s</b>", pa_channel_position_to_pretty_string(map.map[n]));
        addChannel(t);
    }

    g_assert(channels.size() == map.channels);

    levels = NULL;
    display_timeout_signal_connection = Glib::signal_timeout().connect(sigc::mem_fun(*this, &MainWindow::on_display_timeout), 40);
    calc_timeout_signal_connection = Glib::signal_timeout().connect(sigc::mem_fun(*this, &MainWindow::on_calc_timeout), 100);

    show_all();
}

MainWindow::~MainWindow() {
    while (channels.size() > 0) {
        ChannelInfo *i = channels.back();
        channels.pop_back();
        delete i;
    }

    while (levelQueue.size() > 0) {
        LevelInfo *i = levelQueue.back();
        levelQueue.pop_back();
        delete i;
    }

    if (levels)
        delete[] levels;

    display_timeout_signal_connection.disconnect();
    calc_timeout_signal_connection.disconnect();
}

bool MainWindow::on_delete_event(GdkEventAny*) {
    Gtk::Main::quit();
    return true;
}

void MainWindow::addChannel(const Glib::ustring &l) {
    channels.push_back(new ChannelInfo(*this, l));
}

MainWindow::ChannelInfo::ChannelInfo(MainWindow &w, const Glib::ustring &l) {
    label = Gtk::manage(new Gtk::Label(l, 1.0, 0.5));
    label->set_markup(l);

    progress = Gtk::manage(new Gtk::ProgressBar());
    progress->set_fraction(0);

    guint row = w.channels.size()*2;
    w.table.resize(row+1, 2);
    w.table.attach(*label, 0, 1, row, row+1, Gtk::FILL, (Gtk::AttachOptions) 0);
    w.table.attach(*progress, 1, 2, row, row+1, Gtk::EXPAND|Gtk::FILL, (Gtk::AttachOptions) 0);
#ifdef DECIBEL
    // TODO add the dB meter
#endif
}

void MainWindow::pushData(const float *d, unsigned samples) {
    unsigned nchan = channels.size();

    if (!levels) {
        levels = new float[nchan];
        for (unsigned c = 0; c < nchan; c++)
            levels[c] = 0;
    }

    while (samples >= nchan) {

        for (unsigned c = 0; c < nchan; c++) {
            float v = fabs(d[c]);
            if (v > levels[c])
                levels[c] = v;
        }

        d += nchan;
        samples -= nchan;
    }
}

void MainWindow::showLevels(const LevelInfo &i) {
    unsigned nchan = channels.size();

    g_assert(i.levels);

    for (unsigned n = 0; n < nchan; n++) {
        double level;
        ChannelInfo *c = channels[n];

        level = i.levels[n];

#if defined(LOGARITHMIC)
        level = log10(level*9+1);
#elif defined(DECIBEL)
	level = 20*log10(fabs(level));
	level = (level+100)/120;	// show -100dB..+20dB
#endif

	level = level < 0 ? 0 : (level > 1 ? 1 : level);
        c->progress->set_fraction(level);
    }

}

#define DECAY_LEVEL (0.005)

void MainWindow::decayLevels() {
    unsigned nchan = channels.size();

    for (unsigned n = 0; n < nchan; n++) {
        double level;

        ChannelInfo *c = channels[n];

        level = c->progress->get_fraction();

        if (level <= 0)
            continue;

        level = level > DECAY_LEVEL ? level - DECAY_LEVEL : 0;
        c->progress->set_fraction(level);
    }
}

bool MainWindow::on_display_timeout() {
    LevelInfo *i = NULL;

    if (levelQueue.empty()) {
        decayLevels();
        return true;
    }

    while (levelQueue.size() > 0) {
        if (i)
            delete i;

        i = levelQueue.back();
        levelQueue.pop_back();

        if (!i->elapsed())
            break;
    }

    if (i) {
        showLevels(*i);
        delete i;
    }

    return true;
}

bool MainWindow::on_calc_timeout() {
    if (levels) {
        levelQueue.push_front(new LevelInfo(levels, latency));
        levels = NULL;
    }

    return true;
}

void MainWindow::updateLatency(pa_usec_t l) {
    latency = l;
}

static void timeval_add_usec(struct timeval *tv, pa_usec_t v) {
    uint32_t sec = v/1000000;
    tv->tv_sec += sec;
    v -= sec*1000000;

    tv->tv_usec += v;

    while (tv->tv_usec >= 1000000) {
        tv->tv_sec++;
        tv->tv_usec -= 1000000;
    }
}

MainWindow::LevelInfo::LevelInfo(float *l, pa_usec_t latency) {
    levels = l;
    gettimeofday(&tv, NULL);
    timeval_add_usec(&tv, latency);
}

MainWindow::LevelInfo::~LevelInfo() {
    delete[] levels;
}

bool MainWindow::LevelInfo::elapsed() {
    struct timeval now;
    gettimeofday(&now, NULL);

    if (now.tv_sec != tv.tv_sec)
        return now.tv_sec > tv.tv_sec;

    return now.tv_usec >= tv.tv_usec;
}

static MainWindow *mainWindow = NULL;
static pa_context *context = NULL;
static pa_stream *stream = NULL;
static char* device_name = NULL;
static char* device_description = NULL;
static enum {
    PLAYBACK,
    RECORD
} mode = PLAYBACK;


void show_error(const char *txt, bool show_pa_error = true) {
    char buf[256];

    if (show_pa_error)
        snprintf(buf, sizeof(buf), "%s: %s", txt, pa_strerror(pa_context_errno(context)));

    Gtk::MessageDialog dialog(show_pa_error ? buf : txt, false, Gtk::MESSAGE_ERROR, Gtk::BUTTONS_CLOSE, true);
    dialog.run();

    Gtk::Main::quit();
}

static void stream_update_timing_info_callback(pa_stream *s, int success, void *) {
    pa_usec_t t;
    int negative = 0;

    if (!success || pa_stream_get_latency(s, &t, &negative) < 0) {
        show_error("Failed to get latency information");
        return;
    }

    if (!mainWindow)
        return;

    mainWindow->updateLatency(negative ? 0 : t);
}

static gboolean latency_func(gpointer) {
    pa_operation *o;

    if (!stream)
        return false;

    if (!(o = pa_stream_update_timing_info(stream, stream_update_timing_info_callback, NULL)))
        g_message("pa_stream_update_timing_info() failed: %s", pa_strerror(pa_context_errno(context)));
    else
        pa_operation_unref(o);

    return true;
}

static void stream_read_callback(pa_stream *s, size_t l, void *) {
    const void *p;
    g_assert(mainWindow);

    if (pa_stream_peek(s, &p, &l) < 0) {
        g_message("pa_stream_peek() failed: %s", pa_strerror(pa_context_errno(context)));
        return;
    }

    mainWindow->pushData((const float*) p, l/sizeof(float));

    pa_stream_drop(s);
}

static void stream_state_callback(pa_stream *s, void *) {
    switch (pa_stream_get_state(s)) {
        case PA_STREAM_UNCONNECTED:
        case PA_STREAM_CREATING:
            break;

        case PA_STREAM_READY:
            g_assert(!mainWindow);
            mainWindow = new MainWindow(*pa_stream_get_channel_map(s), device_name, device_description);

            g_timeout_add(100, latency_func, NULL);
            pa_operation_unref(pa_stream_update_timing_info(stream, stream_update_timing_info_callback, NULL));
            break;

        case PA_STREAM_FAILED:
            show_error("Connection failed");
            break;

        case PA_STREAM_TERMINATED:
            Gtk::Main::quit();
    }
}

static void create_stream(const char *name, const char *description, const pa_sample_spec &ss, const pa_channel_map &cmap) {
    char t[256];
    pa_sample_spec nss;

    g_free(device_name);
    device_name = g_strdup(name);
    g_free(device_description);
    device_description = g_strdup(description);

    nss.format = PA_SAMPLE_FLOAT32;
    nss.rate = ss.rate;
    nss.channels = ss.channels;

    g_message("Using sample format: %s", pa_sample_spec_snprint(t, sizeof(t), &nss));
    g_message("Using channel map: %s", pa_channel_map_snprint(t, sizeof(t), &cmap));

    stream = pa_stream_new(context, "PulseAudio Volume Meter", &nss, &cmap);
    pa_stream_set_state_callback(stream, stream_state_callback, NULL);
    pa_stream_set_read_callback(stream, stream_read_callback, NULL);
    pa_stream_connect_record(stream, name, NULL, (enum pa_stream_flags) 0);
}

static void context_get_source_info_callback(pa_context *, const pa_source_info *si, int is_last, void *) {
    if (is_last < 0) {
        show_error("Failed to get source information");
        return;
    }

    if (!si)
        return;

    create_stream(si->name, si->description, si->sample_spec, si->channel_map);
}

static void context_get_sink_info_callback(pa_context *, const pa_sink_info *si, int is_last, void *) {
    if (is_last < 0) {
        show_error("Failed to get sink information");
        return;
    }

    if (!si)
        return;

    create_stream(si->monitor_source_name, si->description, si->sample_spec, si->channel_map);
}

static void context_get_server_info_callback(pa_context *c, const pa_server_info*si, void *) {
    if (!si) {
        show_error("Failed to get server information");
        return;
    }

    if (mode == PLAYBACK) {

        if (!si->default_sink_name) {
            show_error("No default sink set.", false);
            return;
        }

        pa_operation_unref(pa_context_get_sink_info_by_name(c, si->default_sink_name, context_get_sink_info_callback, NULL));

    } else if (mode == RECORD) {

        if (!si->default_source_name) {
            show_error("No default source set.", false);
            return;
        }

        pa_operation_unref(pa_context_get_source_info_by_name(c, si->default_source_name, context_get_source_info_callback, NULL));
    }
}

static void context_state_callback(pa_context *c, void *) {
    switch (pa_context_get_state(c)) {
        case PA_CONTEXT_UNCONNECTED:
        case PA_CONTEXT_CONNECTING:
        case PA_CONTEXT_AUTHORIZING:
        case PA_CONTEXT_SETTING_NAME:
            break;

        case PA_CONTEXT_READY:
            g_assert(!stream);

            if (device_name && mode == RECORD)
                pa_operation_unref(pa_context_get_source_info_by_name(c, device_name, context_get_source_info_callback, NULL));
            else if (device_name && mode == PLAYBACK)
                pa_operation_unref(pa_context_get_sink_info_by_name(c, device_name, context_get_sink_info_callback, NULL));
            else
                pa_operation_unref(pa_context_get_server_info(c, context_get_server_info_callback, NULL));

            break;

        case PA_CONTEXT_FAILED:
            show_error("Connection failed");
            break;

        case PA_CONTEXT_TERMINATED:
            Gtk::Main::quit();
    }
}

int main(int argc, char *argv[]) {
    pa_glib_mainloop *m;
    bool record = false;

    signal(SIGPIPE, SIG_IGN);

    try {
        Glib::OptionGroup og("PulseAudio Volume Meter", "Control the volume of your PulseAudio Sound Server");

        Glib::OptionEntry oe;
        oe.set_long_name("record");
        oe.set_description("Show Record Levels");
        og.add_entry(oe, record);

        Glib::OptionContext oc;
        oc.set_main_group(og);

        Gtk::Main kit(argc, argv, oc);

        mode = record ? RECORD : PLAYBACK;

        g_message("Starting in %s mode.", mode == RECORD ? "record" : "playback");

        /* Rather ugly and incomplete */
        if (argc > 1)
            device_name = g_strdup(argv[1]) ;
        else {
            char *e;
            if ((e = getenv(mode == RECORD ? "PULSE_SOURCE" : "PULSE_SINK")))
                device_name = g_strdup(e);
        }

        if (device_name)
            g_message("Using device '%s'", device_name);

        m = pa_glib_mainloop_new(g_main_context_default());
        g_assert(m);

        context = pa_context_new(pa_glib_mainloop_get_api(m), "PulseAudio Volume Meter");
        g_assert(context);

        pa_context_set_state_callback(context, context_state_callback, NULL);
        pa_context_connect(context, NULL, PA_CONTEXT_NOAUTOSPAWN, NULL);

        Gtk::Main::run();

        if (stream)
            pa_stream_unref(stream);
        if (context)
            pa_context_unref(context);

        if (mainWindow)
            delete mainWindow;

        if(device_name)
            g_free(device_name);

        pa_glib_mainloop_free(m);

    } catch (Glib::OptionError) {
        g_message("Bad parameter");
    }

    return 0;
}
